engine/table/src/main/java/io/deephaven/engine/table/impl/MultiJoinTableImpl.java

```diff
@@ -0,0 +1,671 @@
+                if (jd.inputTable().isRefreshing()) {
+                    final QueryTable input = (QueryTable) jd.inputTable();
+                    final ColumnSource<?>[] keySources = getKeySources(jd);
+
+
+                    final String[] matchNames =
+                            Arrays.stream(jd.columnsToMatch()).map(v -> v.right().name()).toArray(String[]::new);
+                    final ModifiedColumnSet sourceKeyModifiedColumnSet =
+                            input.newModifiedColumnSet(matchNames);
+
+                    final String[] addNames =
+                            Arrays.stream(jd.columnsToAdd()).map(v -> v.newColumn().name()).toArray(String[]::new);
+                    resultModifiedColumnSet[ii] =
+                            result.newModifiedColumnSet(addNames);
+                    final ModifiedColumnSet.Transformer transformer =
+                            input.newModifiedColumnSetTransformer(result, addNames);
+
+                    final MultiJoinListenerRecorder listenerRecorder =
+                            new MultiJoinListenerRecorder("multiJoin(" + ii + ")", input, result, keySources,
+                                    sourceKeyModifiedColumnSet, transformer, ii);
+                    listenerRecorder.setMergedListener(mergedListener);
+                    input.addUpdateListener(listenerRecorder);
+                    listenerRecorders.add(listenerRecorder);
+                }
+            }
+
+            result.addParentReference(mergedListener);
+        }
+
+        return result;
+    }
+
+    @NotNull
+    private static MultiJoinInput maybePopulateColumnsToAdd(
+            @NotNull final Collection<String> keyColumnNames,
+            @NotNull final MultiJoinInput[] multiJoinInputs, final int tableNumber) {
+        MultiJoinInput joinInput = multiJoinInputs[tableNumber];
+        JoinAddition[] columnsToAdd = joinInput.columnsToAdd();
+        if (columnsToAdd.length == 0) {
+            // create them on the fly from the table
+            final JoinAddition[] newColumnsToAdd = joinInput.inputTable().getDefinition().getColumnNames().stream()
+                    .filter(cn -> !keyColumnNames.contains(cn)).map(JoinAddition::parse)
+                    .toArray(JoinAddition[]::new);
+            joinInput = multiJoinInputs[tableNumber] =
+                    MultiJoinInput.of(joinInput.inputTable(),
+                            joinInput.columnsToMatch(),
+                            newColumnsToAdd);
+        }
+        return joinInput;
+    }
+
+    private static Table doMultiJoinZeroKey(@NotNull final MultiJoinInput... multiJoinInputs) {
+        final MultiJoinInput[] useJoinInputs = Arrays.copyOf(multiJoinInputs, multiJoinInputs.length);
+        final TObjectIntHashMap<String> usedColumns =
+                new TObjectIntHashMap<>(multiJoinInputs[0].columnsToAdd().length, 0.5f, -1);
+
+        for (int jj = 0; jj < multiJoinInputs.length; ++jj) {
```

Comment:
I wonder if we could implement this loop only once, shared between multi-key and no-key.